File: CodeChef April20A/fctre_1.py
#codechef factor tree problem #code

import math

mod=(10**9)+7

'''Function for calculating prime factors and adding the count of prime numbers in dictionary '''

def calc(path,a):

#print(path)
dic={}
for i in range(len(path)):
    dic=primeFactors(a[path[i]-1],dic)
value=list(dic.values())
ans=1
for i in range(len(value)):
    ans*=(value[i]+1)
    ans=ans%mod
print(ans)
def primeFactors(n,dic):

while n % 2 == 0:
    if 2 not in dic:
        dic[2]=1
    else:
        dic[2]+=1
    n = n / 2
for i in range(3,int(math.sqrt(n))+1,2): 
    while n % i== 0: 
        if i not in dic:
            dic[i]=1
        else:
            dic[i]+=1 
        n = n / i
if n > 2: 
    if n not in dic:
        dic[n]=1
    else:
        dic[n]+=1 
return dic
def addEdge(x, y,v):

v[x].append(y) 
v[y].append(x)  
def DFS(vis, x, y, stack,v,ansflag,a):

stack.append(x) 
if (x == y):
    ansflag=1
    calc(stack,a)
    return 
vis[x] = True
flag = 0
if (len(v[x]) > 0): 
    for j in v[x]: 
        if (vis[j] == False): 
            DFS(vis, j, y, stack,v,ansflag,a) 
            if(ansflag==1):
                flag = 1  
                break
  
if (flag == 0): 
    del stack[-1] 
        
return 
def DFSCall(x, y, n, stack,v,a):

# visited array 
ansflag=0
vis = [0 for i in range(n + 1)] 
if(x!=y):
    x1=DFS(vis, x, y, stack,v,ansflag,a)
else:
    x1=[x]
    calc(x1,a)
return
#main

for _ in range(int(input())):

n = int(input())
v = [[] for i in range(n+1)]

for i in range(n-1):
    a1,b=map(int,input().split())
    addEdge(a1,b,v) 
#print(v)
a=list(map(int,input().split()))
q=int(input())

for i in range(q):
    x,y=map(int,input().split())
    stack = []
    DFSCall(x,y,n,stack,v,a)
#calling DFS for the path in tree

File: CodeChef April20A/fctre_1.cpp
#include <bits/stdc++.h>
using namespace std;
#define ll long int

const int mxA=1e6+1;
const ll MOD=1e9+7;
int prime[mxA];
ll qwe;

void sieve_modified(){
	memset(prime,0,sizeof(prime));
	prime[0]=prime[1]=1;

	for(int i=2;i*i<mxA;i++){
		if(prime[i]==0){
			for(int j=i*2;j<mxA;j+=i){
				if(prime[j]==0){
					prime[j]=i;
				}
			}
		}
	}
	for(int i=2;i<mxA;i++){	
		if(prime[i]==0){
			prime[i]=i;
		}
	}
}

ll numberofDivisorofProduct(vector<int> vec){
	unordered_map<int,int> prime_count;
	int n=vec.size();
	for(int i=0;i<n;i++){
		int temp=vec[i];
		while(temp!=1){
			prime_count[prime[temp]]++;
			temp=temp/prime[temp];
		}
	}
	ll ans=1;
	for(auto it:prime_count){
		ans=(ans*(it.second+1))%MOD;
	}
	return ans%MOD;
} 

void addEdge(vector<int> v[],int x,int y){
	v[x].push_back(y);
	v[y].push_back(x);
}

void DFS(vector<int> v[],bool vis[],int x,int y,vector<int> stack,int cost[],vector<int> fact){
	stack.push_back(x);
	fact.push_back(cost[x-1]);
	if(x==y){
		qwe=numberofDivisorofProduct(fact);
		return ;
	}
	vis[x]=true;

	int flag=0;
	if(!v[x].empty()){
		for(int j=0;j<v[x].size();j++){
			if(vis[v[x][j]]==false){
				DFS(v,vis,v[x][j],y,stack,cost,fact);
			}
		}
	}
}

ll DFScallUtil(int x,int y,vector<int> v[],int n,vector<int> stack,int cost[]){
	bool visited[n+1];
	memset(visited,false,sizeof(visited));
	std::vector<int> fact;
	qwe=0;
	DFS(v,visited,x,y,stack,cost,fact);
	return qwe;
}

void solve(){
	int n;
	cin>>n;
	vector<int> vec[n+1],stack;
	int x,y;
	for(int i=1;i<=n-1;i++){
		cin>>x>>y;
		addEdge(vec,x,y);
	}
	int cost[n];
	for(int i=0;i<n;i++){
		cin>>cost[i];
	}
	int q;
	cin>>q;
	for(int i=0;i<q;i++){
		ll ans123;
		cin>>x>>y;
		/*if(x==y){
			cout<<cost[i];
			continue;
		}*/
		ans123=DFScallUtil(x,y,vec,n+1,stack,cost);
		cout<<ans123%MOD<<"\n";
	}
	return;
}

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(0);

    //freopen("input.txt","r",stdin);
	//freopen("output.txt","w",stdout);

	int t;
	cin>>t;
	sieve_modified();
	while(t--){
		solve();
	}
    return 0;
}